### ISSUE TITLE
Feature request: save/load lyrics in local file

### ISSUE THREAD

**Issue by grwlf** [loskutov/deadbeef-lyricbar#11]
A freature request from me: is it possible to support fetching lyrics from local file? Say, from the file named like the song file, but with 'txt' extension.

I believe that sometimes people need to modify the lyrics, for example:
* they found some error and want to fix it right now
* they want to annotate the text with guitar chords
* they want to translate the song
* etc

So I imagine following modifications:
* A menu button 'save lyrics to file' which indeed saves downloaded lyrics to the file
* Some logic to prefer loading lyrics from the local file if it is exist

**Comment by loskutov** [comment]
I was actually planning to add an option to fetch lyrics using a user-provided script: that could obviously serve your purpose (at least the "load lyrics from the local file" part) with a simple one-liner. However, I'm not sure when I'm ready to release it; as of now, I can suggest the following workarounds:
- As the cached lyrics are stored as regular text files in `~/.cache/deadbeef/lyrics`, you can edit these files or make them symlinks to your `*.txt` lyrics files. Not very convenient, but at least it's something.
- You can embed the lyrics in the file's metadata (media organizers like beets can help you with that). I'd consider it the best way as it works with any more-or-less decent lyrics-shower and doesn't tie you to any particular one.

**Comment by grwlf** [comment]
Both of your suggestions are helpful. I didn't know about beets, thanks. For me, beets seems to be a too complex approach, but I'll investigate how could one use it as a tool. FYI, I also managed to setup shell commands to open corresponding txt in external viewer. It works in general, but there are a number of deadbeef bugs, like not exiting until the viewer is closed. Going to report them soon.

Yes, I think that script-based lyrics loading would be nice!  Then my wish is to support a kind of optional 'edit' button for editing text in a script-dependent way.

I believe that at the end user wants to control how their data is stored, so imo we more like tools which take our storage preferences into account..

**Comment by loskutov** [comment]
I've implemented the external fetchers support; it can be configured inside the plugin configuration, the script execution command can use the whole power of [DeaDBeeF title formatting](DeaDBeeF-Player/deadbeef/wiki/Title-formatting-2.0). However, no updates on editing yet.

**Comment by grwlf** [comment]
Looks great, thanks! 
![image](https://user-images.githubusercontent.com/4477729/67327089-0dc0ab80-f520-11e9-8794-f194406e9d33.png)

**Comment by grwlf** [comment]
For reference: my script loads lyrics from local collection. The plugin is configured as `getlyrics.sh "%artist%" "%title%"` and `~/.local/bin/getlyrics.sh` looks like follows:
```
#!/bin/sh
echo "$0: $@" >&2

LYRICSDIR="$HOME/doc/Music/Chords/"
ARTIST="`echo "$1" | sed 's@^[Tt]he @@'`"
TITLE="$2"

F="$LYRICSDIR/$ARTIST - $TITLE.txt"


if test -f "$F" ; then
  cat "$F"
fi
```

**Comment by grwlf** [comment]
Oh, just discovered, that plugin doesn't re-run script for cached lyrics. It may or may not be a problem, depending of requirements. In my eyes it is more a disadvantage because it is not clear how to update the cache if it happens to exist (but I can clean the cache with my script).

**Comment by grwlf** [comment]
Maybe I found a problem. Load scripts dont work when deadbeef was run not from a terminal. If run from a terminal, everything is fine. Probably some stdin/stdout problem, could you please check?

**Comment by loskutov** [comment]
I couldn't reproduce your issue, but I've rewritten the script spawning routine, please check if it helped.

**Comment by loskutov** [comment]
As of caching, you can right-click your track → Remove Lyrics From Cache.

**Comment by grwlf** [comment]
Well, OK. 2 news:
* Original issue regarding non-terminal runs was on my side. I put my sctip to the folder which wasn't in system PATH, but was in my terminal PATH
* Now, the latest version of plugin (commit 003b85a508dff80910978090b1e67f55bc2ef437) crashs with `terminate called after throwing an instance of 'Gio::Error'`

**Comment by loskutov** [comment]
Fixed in ad0de52.
